fs/fs_zenfs.cc

```diff
@@ -302,8 +319,72 @@ IOStatus ZenFS::RollMetaZone() {
 #ifdef WITH_ZENFS_ASYNC_METAZONE_ROLLOVER
 /* Assumes the files_mtx_ is held */
 IOStatus ZenFS::RollMetaZoneAsync() {
-  return IOStatus::IOError("Calling experimental function "
-                           "ZenFS::RollMetaZoneAsync()!!");
+  ZenMetaLog* new_meta_log;
+  Zone *new_meta_zone, *old_meta_zone;
+  IOStatus s;
+  LatencyHistGuard guard(&zbd_->roll_latency_reporter_);
+  zbd_->roll_qps_reporter_.AddCount(1);
+
+  new_meta_zone = zbd_->AllocateMetaZone();
+
+  if (!new_meta_zone) {
+    assert(false);  // TMP
+    Error(logger_, "Out of metadata zones, we should go to read only now.");
+    return IOStatus::NoSpace("Out of metadata zones");
+  }
+
+  Info(logger_, "Rolling to metazone %d\n", (int)new_meta_zone->GetZoneNr());
+  new_meta_log = new ZenMetaLog(zbd_, new_meta_zone);
+
+  // reserve write pointer to the old metazone to close it later
+  std::unique_ptr<ZenMetaLog> old_meta_log = std::move(meta_log_);
+
+  // shift current write pointer to the new metazone
+  meta_log_.reset(new_meta_log);
+  std::string super_string;
+  superblock_->EncodeTo(&super_string);
+  s = meta_log_->AddRecord(super_string);
+  if (!s.ok()) {
+    Error(logger_,
+          "Could not write super block when rolling to a new meta zone");
+    return IOStatus::IOError("Failed writing a new superblock");
+  }
+
+  BackgroundWorker bg_worker;
+
+  auto task = [&, this](void* arg) {
+    IOStatus s;
+    /* close write for old metazone */
+    old_meta_zone = old_meta_log->GetZone();
+    old_meta_zone->open_for_write_ = false;
+
+    /* Write an end record and finish the meta data zone if there is space left */
+    if (old_meta_zone->GetCapacityLeft()) WriteEndRecord(meta_log_.get());
+    if (old_meta_zone->GetCapacityLeft()) old_meta_zone->Finish();
+
+    // process write snapshot
+    s = WriteSnapshotInSnapshotZone(meta_snapshot_log_.get());
+
+    // reset old meta zones to empty state
+    if (s.ok()) {
+      old_meta_zone->Reset();
+    }
+
+    // TODO: 返回值需要不是int
+    return s.ok();
+  };
+
+  // TODO: arg is a placeholder to satisfy SubmitJob signature
+  int arg = 42;
+  bg_worker.SubmitJob(task, (void*)&arg);
```

Comment:
This is not good-looking... do you think we can remove the args here? I think lambda-based task is enough. @levisonchen

Reply:
We could structurize that, but we also needs arguments for jobs, it's just a change of coding style.